arangod/Graph/Enumerators/YenEnumerator.cpp

```diff
@@ -0,0 +1,380 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2014-2024 ArangoDB GmbH, Cologne, Germany
+/// Copyright 2004-2014 triAGENS GmbH, Cologne, Germany
+///
+/// Licensed under the Business Source License 1.1 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     https://github.com/arangodb/arangodb/blob/devel/LICENSE
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Max Neunhoeffer
+////////////////////////////////////////////////////////////////////////////////
+
+#include "YenEnumerator.h"
+
+#include "Basics/Exceptions.h"
+#include "Basics/debugging.h"
+#include "Basics/system-compiler.h"
+#include "Basics/voc-errors.h"
+
+#include "Futures/Future.h"
+#include "Graph/Options/TwoSidedEnumeratorOptions.h"
+#include "Graph/PathManagement/PathResult.h"
+#include "Graph/PathManagement/PathStore.h"
+#include "Graph/PathManagement/PathStoreTracer.h"
+#include "Graph/PathManagement/PathValidator.h"
+#include "Graph/Providers/ClusterProvider.h"
+#include "Graph/Providers/ProviderTracer.h"
+#include "Graph/Providers/SingleServerProvider.h"
+#include "Graph/Steps/SingleServerProviderStep.h"
+#include "Graph/Types/ValidationResult.h"
+#include "Graph/algorithm-aliases.h"
+
+#include <Logger/LogMacros.h>
+#include <velocypack/Builder.h>
+#include <velocypack/HashedStringRef.h>
+
+using namespace arangodb;
+using namespace arangodb::graph;
+
+template<class ProviderType, class EnumeratorType, bool IsWeighted>
+YenEnumerator<ProviderType, EnumeratorType, IsWeighted>::YenEnumerator(
+    ProviderType&& forwardProvider, ProviderType&& backwardProvider,
+    TwoSidedEnumeratorOptions&& options, PathValidatorOptions validatorOptions,
+    arangodb::ResourceMonitor& resourceMonitor)
+    : _arena(resourceMonitor),
+      _resourceMonitor(resourceMonitor),
+      _totalMemoryUsageHere(0),
+      _isDone(true),
+      _isInitialized(false) {
+  // Yen's algorithm only ever uses the TwoSidedEnumerator here to find
+  // exactly one shortest path:
+  options.setOnlyProduceOnePath(true);
+  options.setPathType(PathType::Type::ShortestPath);
+  _shortestPathEnumerator = std::make_unique<EnumeratorType>(
+      std::move(forwardProvider), std::move(backwardProvider),
+      std::move(options), std::move(validatorOptions), resourceMonitor);
+  _shortestPathEnumerator->setEmitWeight(true);
```

Comment:
It would be possible, but these enumerators are used in extremely many places (see `EnumeratePathsNode::createBlock` and most of them do not use this `setEmitWeight` option to `true`. And since there are multiple template types (`TwoSidedEnumerator` and `WeightedTwoSidedEnumerator`) we would have to change the constructor signature in both and then change the constructor call in a lot of places, everywhere putting the explicit `false` in. I would rather not want to touch so much code.